cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationNotificationSubscriberService.java

```diff
@@ -0,0 +1,353 @@
+    this.transactionRunner = transactionRunner;
+    this.manager = manager;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    List<Service> children = new ArrayList<>();
+    String topicPrefix = cConf.get(Constants.Operation.STATUS_EVENT_TOPIC);
+    int numPartitions = cConf.getInt(Constants.Operation.STATUS_EVENT_NUM_PARTITIONS);
+    IntStream.range(0, numPartitions)
+        .forEach(i -> children.add(createChildService("operation.status." + i, topicPrefix + i)));
+    delegate = new CompositeService(children);
+
+    delegate.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    delegate.stopAndWait();
+  }
+
+  private OperationNotificationSingleTopicSubscriberService createChildService(
+      String name, String topicName) {
+    return new OperationNotificationSingleTopicSubscriberService(
+        messagingService,
+        cConf,
+        metricsCollectionService,
+        statePublisher,
+        transactionRunner,
+        name,
+        topicName,
+        manager);
+  }
+}
+
+
+/**
+ * Service that receives program status notifications from a single topic and persists to the store.
+ * No transactions should be started in any of the overrided methods since they are already wrapped
+ * in a transaction.
+ */
+class OperationNotificationSingleTopicSubscriberService
+    extends AbstractNotificationSubscriberService {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OperationNotificationSubscriberService.class);
+
+  private static final Gson GSON =
+      ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder()).create();
+
+  private final OperationStatePublisher statePublisher;
+  private final Queue<Runnable> tasks;
+  private final CConfiguration cConf;
+
+  private final OperationLifecycleManager manager;
+
+  OperationNotificationSingleTopicSubscriberService(
+      MessagingService messagingService,
+      CConfiguration cConf,
+      MetricsCollectionService metricsCollectionService,
+      OperationStatePublisher statePublisher,
+      TransactionRunner transactionRunner,
+      String name,
+      String topicName,
+      OperationLifecycleManager manager) {
+    super(
+        name,
+        cConf,
+        topicName,
+        cConf.getInt(Constants.Operation.STATUS_EVENT_FETCH_SIZE),
+        cConf.getLong(Constants.Operation.STATUS_EVENT_POLL_DELAY_MILLIS),
+        messagingService,
+        metricsCollectionService,
+        transactionRunner,
+        cConf.getInt(Constants.Operation.STATUS_EVENT_TX_SIZE));
+    this.statePublisher = statePublisher;
+    this.manager = manager;
+    this.tasks = new LinkedList<>();
+    this.cConf = cConf;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    RetryStrategy retryStrategy =
+        RetryStrategies.fromConfiguration(cConf, Constants.Service.RUNTIME_MONITOR_RETRY_PREFIX);
+    Retries.runWithRetries(
+        this::startPendingRuns,
```

Comment:
I have moved this to `OperationRuntimeService`. It will process all active runs during startup.

The same pattern is used in ProgramNotificationSubscriberService so maybe we also want to rethink the approach there also.